drivers: ipm_stm32wb: add workaround for conn_complete

pavlohamov opened the pull request:
Enhanced connection complete message lacks of
the peer address (FF::FF), which leads to pairing/bonding failure.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/19509

Commit:
drivers: ipm_stm32wb: add workaround for conn_complete

Enhanced connection complete message lacks of
the peer address (FF::FF), which leads to pairing/bonding failure.

Fixes: #19509

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1032,10 +1032,20 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
+			bt_addr_str(&evt->peer_rpa), bt_addr_str(&evt->local_rpa));
 
 #if defined(CONFIG_BT_SMP)
+	if (!bt_addr_cmp(&evt->peer_addr.a, BT_ADDR_NONE)) {
```
This extra code should be build-time conditional to the HCI driver that exhibits this issue. You should also add a code comment here explaining it (basically the same as you have in the commit message). We could also entertain the thought of introducing driver quirks (same as Linux HCI drivers have), but I think a Kconfig dependency is enough for now.

Alternatively, could you do the fixup directly inside the HCI driver, to avoid polluting the HW-agnostic host-side code with this kind of workarounds?

joerchan replied:
I think that putting the code in ipm_stm32wb.c might be the best option. That would leave the least amount of clutter I think. Especially in case somebody wants to use the BT_ADDR_NONE address with an IRK.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1032,10 +1032,20 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
 
 #if defined(CONFIG_BT_SMP)
+	if (!bt_addr_cmp(&evt->peer_addr.a, BT_ADDR_NONE)) {
+		if (bt_addr_cmp(&evt->peer_rpa, BT_ADDR_NONE)) {
```
Is the ST firmware actually sometimes giving your BT_ADDR_NONE as RPA? If not, I'd just leave out this secondary check.

pavlohamov replied:
Totally agree with your point. I'm not familiar with the raw HCI packet format, thus it was easier for me to write code for the "parsed" HCI packet. Now I'm investigating which bytes to copy in the driver

joerchan replied:
@pavlohamov Yeah, it is a bit more tricky to add it there.

Your buffer will consist of these:
struct bt_hci_evt_hdr
check if evt == BT_HCI_EVT_LE_META_EVENT
struct bt_hci_evt_le_meta_event
check if subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE
struct bt_hci_evt_le_enh_conn_complete

pavlohamov replied:
@joerchan Thank you for your directions. I've dumped the message content and understand that it could be simply cast'ed to bt_hci_evt_le_enh_conn_complete. Please take a look

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -1032,10 +1032,20 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
+		} else {
+			BT_WARN("Invalid peer address %s", bt_addr_le_str(&evt->peer_addr));
+		}
+	}
```
minor coding style: empty line after this (whenever the next line is not also a `}`)

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,31 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+ */
+		struct bt_hci_evt_le_meta_event *mev;
+		mev = (struct bt_hci_evt_le_meta_event *)hcievt->evtserial.evt.payload;
+		if (mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
```
Check that it is in fact an LE_Meta_Event.
```suggestion
		if (hcievt->evtserial.evt.evtcode == BT_HCI_EVT_LE_META_EVENT &&
		    mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
```

pavlohamov replied:
Make sense. Thanks, Done

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,31 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+		if (mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
+			struct bt_hci_evt_le_enh_conn_complete *evt =
+					(struct bt_hci_evt_le_enh_conn_complete*)(mev + 1);
+			BT_WARN("Invalid peer addr %s", bt_addr_le_str(&evt->peer_addr));
```
Move warning into the if-statement?

pavlohamov replied:
Oups, my bad. Was fixing the 80 characters issue and moved to improper place. Thanks, done

joerchan reviewed (approved):
Looks OK.
You have a few lines that are too long, and there is a requirement to have blank line after variable declarations.

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,33 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
 		}
+/*
+ * https://github.com/zephyrproject-rtos/zephyr/issues/19509
+ * For some reason BLE stack (v 1.2.0) is not reporting the peer address
```
The bug as it appears to me is that: "Unresolved Resolvable Private Addresses (RPA) is reported in the peer_rpa field, and not in the peer address, as it should, when this happens the peer address is set to all FFs"
Might be a better explanation for what you are doing here.

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,33 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+ * For some reason BLE stack (v 1.2.0) is not reporting the peer address
+ * 0A 00 01 08 01 01 FF FF FF FF FF FF 00 00 00 00 00 00 0C AA C5 B3 3D 6B ...
+ * If such message is passed to HCI core than no Secure connection could be
+ * established. Temporary copy RPA to PEER address
```
It's not temporary.  Should update to "Rewrite the event with the RPA in the PEER address field"

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,33 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+ * https://github.com/zephyrproject-rtos/zephyr/issues/19509
+ * For some reason BLE stack (v 1.2.0) is not reporting the peer address
+ * 0A 00 01 08 01 01 FF FF FF FF FF FF 00 00 00 00 00 00 0C AA C5 B3 3D 6B ...
+ * If such message is passed to HCI core than no Secure connection could be
```
Legacy pairing is also affected. Essentially the Pairing procedure will fail.

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -119,8 +119,33 @@ void TM_EvtReceivedCb(TL_EvtPacket_t *hcievt)
+ * If such message is passed to HCI core than no Secure connection could be
+ * established. Temporary copy RPA to PEER address
+ */
+		struct bt_hci_evt_le_meta_event *mev;
```
With all the 80 character limit violations you might consider just moving this into a function.

pavlohamov commented:
@joerchan Thank you for your review. I've updated the description and moved solution to separate function. Added new lines after the variable declaration

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+	if (tev->evtcode == BT_HCI_EVT_LE_META_EVENT &&
+			mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
+		struct bt_hci_evt_le_enh_conn_complete *evt =
+				(struct bt_hci_evt_le_enh_conn_complete *)(mev + 1);
```
Are you sure this does the right thing? Doesn't the `(mev + 1)` in translate to `mev + 1 * sizeof(*mev)`? I.e. you need a `u8_t *` typecast somewhere there, I think, e.g. :  `(struct bt_hci_evt_le_enh_conn_complete *)((u8_t *)mev + 1)`

jhedberg replied:
Which actually begs the question, did you test this? 😃

pavlohamov replied:
Yeap, i've tested it. And i was expecting here see the same: pointer incrimination by size of the struct, not simply as the integer. Without any modification to pointer calculation
`BT_WARN("%p %p", mev, evt);`
```
[00:00:09.560,900] <dbg> hci_ipm.TM_EvtReceivedCb: EVT: hcievt->evtserial.evt.evtcode: 0x3e
[00:00:09.560,900] <wrn> hci_ipm: 0x20038437 0x20038438
[00:00:09.561,100] <wrn> hci_ipm: Invalid peer addr ff:ff:ff:ff:ff:ff (random)
```
Seems like GCC is too smart.
But you right that it's making sense to cast pointer to u8_t* before addition.
Updated.

pavlohamov replied:
> 
> 
> Which actually begs the question, did you test this? 😃

:) Yep, I've tested it. And there is some "pointer magic"

joerchan replied:
sizeof(struct bt_hci_evt_le_meta_event) is 1 i think. But isn't that correct, you want to move past the struct and get the bytes after that?

jhedberg replied:
> sizeof(struct bt_hci_evt_le_meta_event) is 1 i think. 

Indeed it is, and that explains why the pointer arithmetic gave the right result.

> But isn't that correct, you want to move past the struct and get the bytes after that?

Yes, you're right, I misunderstood the intent of the code since this kind of pointer arithmetic bugs happen quite easily with the special rules C has.

pavlohamov replied:
Got your point. Moving by hard-coded 1 byte offset is bad idea. Updated and added sizeof instead of hard code.

pavlohamov commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+{
+/*
+ * https://github.com/zephyrproject-rtos/zephyr/issues/19509
+ * Tested on BLE stack (v1.2.0)
```
Sure. Thank you for noticing. My comment is a bit confusing. Done

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+				(struct bt_hci_evt_le_enh_conn_complete *)ptr;
+		if (!bt_addr_cmp(&evt->peer_addr.a, BT_ADDR_NONE)) {
+			BT_WARN("Invalid peer addr %s", bt_addr_le_str(&evt->peer_addr));
+			if (bt_addr_cmp(&evt->peer_rpa, BT_ADDR_NONE)) {
```
This seems like an unnecessary branch if you haven't actually seen the controller report an invalid peer_rpa.

pavlohamov replied:
Have never seen station like that. Idea was just to be sure that we do not copy some new garbage. You right - keep i simple. Done

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+			mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
+		struct bt_hci_evt_le_enh_conn_complete *evt =
+				(struct bt_hci_evt_le_enh_conn_complete *)ptr;
+		if (!bt_addr_cmp(&evt->peer_addr.a, BT_ADDR_NONE)) {
```
Wouldn't `bt_addr_le_cmp(&evt->peer_addr, BT_ADDR_LE_NONE)` work?

joerchan replied:
LE_NONE doesn't work because of the address type is public, but this one is random.

jhedberg replied:
Ok, fair point - just ignore this comment then

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+static void tryfix_event(TL_Evt_t *tev)
+{
+/*
+ * https://github.com/zephyrproject-rtos/zephyr/issues/19509
```
Normally code comments use the same indentation level as the code itself, meaning this comment block should be indented by one extra tab.

jhedberg replied:
Alternatively, move the block right above the function (probably simpler since you've already got the text pre-formatted for the the full width)

pavlohamov replied:
Moved right above the function. Thanks.

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+	u8_t *ptr = ((u8_t *)mev + (sizeof(*mev)));
+
+	if (tev->evtcode == BT_HCI_EVT_LE_META_EVENT &&
+			mev->subevent == BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
```
This should be aligned with the opening ( on the first line

pavlohamov replied:
Oh, it even looks better. Thanks. Done.

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+	u8_t *ptr = ((u8_t *)mev + (sizeof(*mev)));
+
+	if (tev->evtcode != BT_HCI_EVT_LE_META_EVENT ||
+		mev->subevent != BT_HCI_EVT_LE_ENH_CONN_COMPLETE) {
```
Still not quite right :) It should be one tab and then 4 spaces to align with the same position as `tevt->evtcode` on the first line.

pavlohamov replied:
Ah, you are faster than I'm! just was seeking for example in the code. Done

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,37 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+{
+	struct bt_hci_evt_le_meta_event *mev =
+			(struct bt_hci_evt_le_meta_event *)&tev->payload;
+	u8_t *ptr = ((u8_t *)mev + (sizeof(*mev)));
```
This `ptr` variable seems a bit unnecessary - how come you don't just assign directly to `evt`? Even if you did have an intermediate variable wouldn't `void *` be better since it doesn't require explicit typecasting? (I'd still go for just dropping this variable however)

jhedberg replied:
@pavlohamov btw, thanks for your patience with the reviews. Once this last small thing is sorted I think we can go and merge this.

pavlohamov replied:
It's just a matter of fitting in one line. for one cast from ptr is much more clear than 
```
struct bt_hci_evt_le_enh_conn_complete *evt =
		(struct bt_hci_evt_le_enh_conn_complete *)((u8_t *)mev +
		(sizeof(*mev)));
```
No problem. Thank you for the pedantic review. During review I've learned couple of thing about zephyr HCI core. Now I understand the internals of HCI packet. at least one =)

jhedberg replied:
You could do a `(void *)` cast to make the line shorter. It's equally valid (as in, the compiler won't complain) as the exact type on the left side of `=`

pavlohamov replied:
Make sense. Now two extra lines are saved. Thanks,

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -96,6 +96,35 @@ static void syscmd_status_not(SHCI_TL_CmdStatus_t status)
+	}
+
+	struct bt_hci_evt_le_enh_conn_complete *evt = (void *)((u8_t *)mev +
+			(sizeof(*mev)));
```
I'm going to approve now, since I feel like I've exceeded my "being pedantic" quota 😄 , but if you split it like this then you need to align it again with the right ( on the previous line since the second line continues something inside (). Alternatively, and what might be more readable, is to just split after the `=` so you have the full pointer manipulation on a single line.

pavlohamov replied:
Got it. Done. On the one line ok

joerchan commented:
@erwango Your comments have been addressed. Does it look ok?
